graphrag/index/flows/create_final_embeddings.py

```diff
@@ -0,0 +1,211 @@
+                filename="create_final_entities_description_embeddings",
+                callbacks=callbacks,
+                cache=cache,
+                storage=storage,
+                base_text_embed=base_text_embed,
+                message="Entities embeddings (description)",
+            )
+
+        # Embed title in community reports table
+        if community_title_embedding in embedded_fields and isinstance(
+            community_reports_embeddings, pd.DataFrame
+        ):
+            await run_embeddings(
+                data=community_reports_embeddings.copy(),
+                column_to_embed="title",
+                filename="create_final_community_reports_title_embeddings",
+                callbacks=callbacks,
+                cache=cache,
+                storage=storage,
+                base_text_embed=base_text_embed,
+                message="Community reports embeddings (title)",
+            )
+
+        # Embed summary in community reports table
+        if community_summary_embedding in embedded_fields and isinstance(
+            community_reports_embeddings, pd.DataFrame
+        ):
+            await run_embeddings(
+                data=community_reports_embeddings.copy(),
+                column_to_embed="summary",
+                filename="create_final_community_reports_summary_embeddings",
+                callbacks=callbacks,
+                cache=cache,
+                storage=storage,
+                base_text_embed=base_text_embed,
+                message="Community reports embeddings (summary)",
+            )
+
+        # Embed full_content in community reports table
+        if community_full_content_embedding in embedded_fields and isinstance(
+            community_reports_embeddings, pd.DataFrame
+        ):
+            await run_embeddings(
+                data=community_reports_embeddings.copy(),
+                column_to_embed="full_content",
+                filename="create_final_community_reports_full_content_embeddings",
+                callbacks=callbacks,
+                cache=cache,
+                storage=storage,
+                base_text_embed=base_text_embed,
+                message="Community reports embeddings (full_content)",
+            )
+
+    return pd.DataFrame()
+
+
+async def run_embeddings(
+    data: pd.DataFrame,
+    column_to_embed: str,
+    filename: str,
+    callbacks: VerbCallbacks,
+    cache: PipelineCache,
+    storage: PipelineStorage,
+    base_text_embed: dict | None = None,
+    message: str | None = None,
+) -> None:
+    """All the steps to generate single embedding."""
+    if base_text_embed:
+        log.info("%s", message)
+        data["embedding"] = await embed_text(
+            data,
+            callbacks,
+            cache,
+            column=column_to_embed,
+            strategy=base_text_embed["strategy"],
+        )
+
+        data = cast(pd.DataFrame, data[["id", "embedding"]])
+
+    await snapshot(
```

Comment:
I don't think running embedding should also store the file. These are separate concerns.

Reply:
remember that now the embeddings are not in memory at all, the full process ends at disk write so I believe it should be in the same function.

Reply:
The disagree. The purpose of the function is to run embeddings and store it on a column. Why mis it with I/O operations in the same context?

Reply:
because this column won't be saved in memory at all, the embeddings is just an I/O operation now.

Reply:
Same as below, if we use slicing that ensures df retrieval we can avoid type checks and casting

Reply:
What I mean is in this function we want to return data, and then call snapshot in another place after calling run_embeddings. So run_embeddings calculates embedding fields without side effect like saving to disk which should be the concern of this method or wouldn't be expected.

Reply:
@natoverse what do you think?

Reply:
I agree with both points. I think Alonso is right that it feels weird to bundle these two concerns into a single function. However, because the entire point of this workflow is to run + save the embeddings, this function more serves to reduce duplicate code than anything, so it's not bothering me much. A couple thoughts:

- It would help to rename it something like "run_and_snapshot_embeddings" if we keep them bundled
- It also might help to put `_` at the front to indicate it is an internal function for the module
- Can we measure Gaudy's point? If the function call indirection unloads the temporary dataframe from the call stack and reduces peak memory, this definitely seems worth bundling into one function.
- To the last point: up-front in the flow is the creation of a bunch of selected dataframes, these could each be put after the if statement for their associated embeddings to co-locate the logic, and maybe save memory